File: tests/smoke/__implicit_checklist_routing_smoke.js
// Implicit checklist mutation routing — no literal "checklist" required.
// Run: node tests/smoke/__implicit_checklist_routing_smoke.js

"use strict";
const fs = require("fs");
const path = require("path");
const vm = require("vm");

const CHECKLIST_JS = fs.readFileSync(path.resolve(__dirname, "../../workmode/checklist.js"), "utf8");
const APP_JS = fs.readFileSync(path.resolve(__dirname, "../../app/app.js"), "utf8");

const blockStart = CHECKLIST_JS.indexOf("const CHECKLIST_IMPLICIT_BLOCK_RES");
const blockEnd = CHECKLIST_JS.indexOf("try { window.detectImplicitChecklistMutation");
const implicitBlock = CHECKLIST_JS.slice(blockStart, blockEnd);

const compoundStart = APP_JS.indexOf("const _WMC_PANEL_FAMILY_RE =");
const compoundEnd = APP_JS.indexOf("try { window.detectCompoundActionFamilies");
const compoundBlock = APP_JS.slice(compoundStart, compoundEnd);

const ctx = { window: {}, module: { exports: {} }, exports: {}, console };
vm.createContext(ctx);
vm.runInContext(
  `"use strict";
  const CHECKLIST_NOUN_RE = /\\b(?:checklist|check\\s+list|to-?do(?:\\s+list)?|item|items|task|tasks)\\b/i;
  const CHECKLIST_NON_OBJECT_NOUN_RE = /\\b(?:paragraph|sentence|argument|example|essay|panel|timer|volume|minute|minutes|detail|evidence)\\b/i;
  const CHECKLIST_UNCOMPLETE_VERB_RE = /\\b(?:uncheck|undo\\s+complete|mark\\b.*\\bincomplete\\b)/i;
  ${implicitBlock}
  function isLikelyWorkChecklistEditIntent(text) {
    const implicit = detectImplicitChecklistMutation(text);
    return Boolean(implicit?.detected && implicit.count >= 1);
  }
  ${compoundBlock}
  module.exports = {
    detectImplicitChecklistMutation,
    isLikelyWorkChecklistEditIntent,
    detectCompoundActionFamilies,
  };
  `,
  ctx,
  { filename: "implicit-checklist-routing.js" }
);

const { detectImplicitChecklistMutation, isLikelyWorkChecklistEditIntent, detectCompoundActionFamilies } =
  ctx.module.exports;

let pass = 0;
let fail = 0;
function ok(cond, name, detail = "") {
  if (cond) {
    pass++;
    console.log(`  PASS  ${name}`);
  } else {
    fail++;
    console.log(`  FAIL  ${name}${detail ? " — " + detail : ""}`);
  }
}

console.log("\n-- 1. Compound add + mark without checklist word --");
{
  const text = "can you add stat homework and mark milk and eggs complete?";
  const implicit = detectImplicitChecklistMutation(text);
  ok(implicit.detected === true, "implicit mutation detected");
  ok(implicit.count === 2, `mutation count=${implicit.count}`);
  ok(isLikelyWorkChecklistEditIntent(text) === true, "reasoning gate skips checklist edit");
  const compound = detectCompoundActionFamilies(text);
  ok(compound.isCompound === true, "compound defers to backend planner");
}

console.log("\n-- 2. Explicit checklist wording still works --");
{
  const text =
    "can you add stat homework and mark milk and eggs complete in the checklist?";
  ok(detectImplicitChecklistMutation(text).detected === true, "implicit still detected");
  ok(isLikelyWorkChecklistEditIntent(text) === true, "checklist edit intent");
}

console.log("\n-- 3-5. Single implicit mutations --");
ok(detectImplicitChecklistMutation("add milk and eggs").detected === true, "add milk and eggs");
ok(
  detectImplicitChecklistMutation("mark milk and eggs complete").detected === true,
  "mark milk and eggs complete"
);
ok(detectImplicitChecklistMutation("remove milk and eggs").detected === true, "remove milk and eggs");

console.log("\n-- 6-10. Negative routing guards --");
ok(detectImplicitChecklistMutation("add 3 minutes to the timer").detected === false, "timer extend");
ok(detectImplicitChecklistMutation("remove panel 4").detected === false, "panel close");
ok(
  detectImplicitChecklistMutation("add more detail to the explanation").detected === false,
  "reasoning refinement add detail"
);
ok(
  detectImplicitChecklistMutation("add evidence to my essay in panel 3").detected === false,
  "panel reasoning add evidence"
);
ok(detectImplicitChecklistMutation("complete this explanation").detected === false, "complete explanation");

console.log(`\nTotal: ${pass + fail}  PASS=${pass}  FAIL=${fail}`);
process.exit(fail ? 1 : 0);

File: tests/smoke/__implicit_checklist_routing_smoke.py
"""Implicit checklist routing smoke tests."""
from __future__ import annotations

import os
import sys

HERE = os.path.dirname(os.path.abspath(__file__))
ROOT = os.path.abspath(os.path.join(HERE, "..", ".."))
if ROOT not in sys.path:
    sys.path.insert(0, ROOT)

from actions.checklist import is_checklist_action_request  # noqa: E402
from actions.multi_action_planner import plan_user_actions, should_trigger_planner  # noqa: E402


def _types(plan: dict) -> list[str]:
    return [a.get("type") or "" for a in plan.get("actions") or []]


def main() -> int:
    fail = 0

    text = "can you add stat homework and mark milk and eggs complete?"
    triggered, reason = should_trigger_planner(text)
    plan = plan_user_actions(text)
    types = _types(plan)
    if not triggered:
        print("FAIL should_trigger_planner")
        fail += 1
    else:
        print(f"PASS should_trigger_planner reason={reason}")
    if types != ["checklist.add", "checklist.complete"]:
        print(f"FAIL plan types={types}")
        fail += 1
    else:
        print("PASS plan types checklist.add + checklist.complete")
    if is_checklist_action_request(text) is None:
        print("FAIL is_checklist_action_request")
        fail += 1
    else:
        print("PASS is_checklist_action_request")

    explicit = text + " in the checklist"
    if _types(plan_user_actions(explicit)) != ["checklist.add", "checklist.complete"]:
        print("FAIL explicit checklist plan")
        fail += 1
    else:
        print("PASS explicit checklist plan")

    negatives = [
        "add 3 minutes to the timer",
        "remove panel 4",
        "add more detail to the explanation",
        "add evidence to my essay in panel 3",
        "complete this explanation",
    ]
    for neg in negatives:
        if is_checklist_action_request(neg) is not None:
            print(f"FAIL negative routed to checklist: {neg!r}")
            fail += 1
        else:
            print(f"PASS negative blocked: {neg!r}")

    print(f"\nFAIL={fail}")
    return 1 if fail else 0


if __name__ == "__main__":
    raise SystemExit(main())
